text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+This is generally quite inefficient but we are also running into limitations in the workers.  For instance
+we rely exclusively on memcache for caches which means that we are limited by our cache size limitations.
+Anything above the cache size is not processable which can cause a bad user experience for users depending
+on very large source maps.
+
+Symbolicator has a file system based caching system with GCS backing for rapid synching of new symbolicator
+instances we could leverage.  Additionally symbolicator has a better HTTP fetching system than Sentry as
+it's able to concurrencly fetch whereas the Sentry codebase is not.
+
+Future feature improvements on the source map side might also demand more complexity in the processing side
+which we are not currently considering given already existing scalability concerns on the existing workers.
+
+## Python Monolith
+
+These are changes to break up the monolith.  The goals here are largely that different teams can work largely
+uninterrupted of each other, even on the monolith.  For instance in an ideal scenario changes to an integration
+platform code do not necessarily have to run the entire Sentry testsuite on every commit.  Likewise UI code
+ideally does not build the entirety of the Sentry UI codebase.
+
+### Import Compartmentalization
+
+As a first start imports in the Sentry codebase should be compartmentalized as much as possible.  Today we have
+some catch all modules like `sentry.models` which import everything.  This means that it's hard for as to track
+the actual dependencies from different pieces of code and also hides away circular imports.  One of the results of
+this is that if you run a test against a single pytest module, you still pull in the entire codebase which already
+takes a couple of seconds.  It also means that it's harder for test infrastructure code to analyze the minimal set
+of dependencies that might need testing.
+
+Likewise code in `sentry.utils` should most likely no longer import models etc.  Some of that code might event
+move out of `sentry` entirely.
+
+### Move Code from Getsentry to Senty
+
+Currently quite a bit of code lives in `getsentry` for reasons of assumed convenience.  However this has created
+the situation that it is quite easy to ship regressions because the changes in `getsentry` were not considered.
+There is likely a whole range of code that does not need to live in `getsentry` and moving it to `sentry` would
+make the testsuite more reliable, faster to run and catch more regressions early.
+
+### Remove Pickle in Database
+
+We still use pickled models in database code quite extensively.  This reliance on pickle makes it harder than
+necessary to read and write from this outside of the main Python codebase.  Moving this to JSON will simplify the
+handling of moving such code into environments where other code wants to access it.  We already require depickling
+in the data pipeline for such cases.
+
+### Phase out Buffers where Possible
```

Comment:
honestly i think buffers is conceptually a good idea to keep. it provides batching for a subset of operations in your task/function without forcing you to give up isolation of tasks/functions. I instead wish we made buffers more scalable under the hood and extended the abstraction to work for actually more things than models.

Reply:
I think buffers is a good concept and it should stay. The way it's built is maybe not ideal (in particular the model and pickle dependency) and particularly given extended outages as a possibility it would be preferrable this to be backed by kafka rather than redis for instance. I don't think everything should build a kafka consumer but a buffer service could be a possibility.